File: llm_harness/stats/llm/llm_stats_stages/scoring.py
"""Score helpers for native Python LLM stats."""

from __future__ import annotations

from statistics import median
from typing import Any

import polars as pl

from ..shared import as_finite_number, as_record


def percentile_rank(values: list[Any], value: Any) -> float | None:
    """Compute a finite-aware aggregate for Scoring LLM stats selection."""
    numeric_value = as_finite_number(value)
    if numeric_value is None:
        return None
    finite_values = [item for item in (as_finite_number(v) for v in values) if item is not None]
    if not finite_values:
        return None
    return pl.DataFrame({"value": finite_values}).select(((pl.col("value") <= numeric_value).sum() / pl.len()) * 100).item()


def mean_of_finite(values: list[float | None]) -> float | None:
    """Compute a finite-aware aggregate for Scoring LLM stats selection."""
    finite_values = [value for value in values if value is not None]
    if not finite_values:
        return None
    return pl.Series("value", finite_values).mean()


def _metric_value(model: dict[str, Any], key: str) -> float | None:
    """Helper for metric value."""
    intelligence = as_record(model.get("intelligence"))
    evaluations = as_record(model.get("evaluations"))
    if key == "omniscience_nonhallucination_rate":
        return as_finite_number(intelligence.get("omniscience_nonhallucination_rate")) or as_finite_number(intelligence.get("omniscience_hallucination_rate"))
    return as_finite_number(intelligence.get(key)) or as_finite_number(evaluations.get(key))


def blended_price_value(cost_like: Any, scoring_config: dict[str, Any]) -> float | None:
    """Helper for blended price value."""
    cost = as_record(cost_like)
    input_cost = as_finite_number(cost.get("input"))
    output_cost = as_finite_number(cost.get("output"))
    weighted_input = as_finite_number(cost.get("weighted_input"))
    weighted_output = as_finite_number(cost.get("weighted_output"))
    cache_read = as_finite_number(cost.get("cache_read"))
    cache_write = as_finite_number(cost.get("cache_write"))
    if input_cost is None or output_cost is None or input_cost <= 0 or output_cost <= 0:
        return None
    input_ratio = scoring_config.get("weighted_price_input_ratio", 0.75)
    output_ratio = scoring_config.get("weighted_price_output_ratio", 0.25)
    if weighted_input is not None or weighted_output is not None:
        effective_input = weighted_input if weighted_input is not None else input_cost
        effective_output = weighted_output if weighted_output is not None else output_cost
        return input_ratio * effective_input + output_ratio * effective_output
    cache_weighted_input = cache_read if cache_read is not None else input_cost
    cache_weighted_output = 0.1 * cache_write + 0.9 * output_cost if cache_write is not None else output_cost
    return input_ratio * (input_ratio * cache_weighted_input + output_ratio * input_cost) + output_ratio * cache_weighted_output


def derive_speed_output_token_anchors(openrouter_speed_by_id: dict[str, dict[str, Any]], scoring_config: dict[str, Any]) -> list[int]:
    """Derive the speed output token anchors."""
    implied_token_usages: list[float] = []
    for speed in openrouter_speed_by_id.values():
        throughput = as_finite_number(speed.get("throughput_tokens_per_second_median"))
        latency = as_finite_number(speed.get("latency_seconds_median"))
        e2e_latency = as_finite_number(speed.get("e2e_latency_seconds_median"))
        if throughput is None or throughput <= 0 or latency is None or e2e_latency is None:
            continue
        generation_seconds = e2e_latency - latency
        if generation_seconds <= 0:
            continue
        implied_token_usages.append(generation_seconds * throughput)
    implied_token_usages.sort()
    default_anchors = list(scoring_config.get("default_speed_output_token_anchors", [200, 500, 1000, 2000, 8000]))
    if not implied_token_usages:
        return default_anchors
    quantiles = [0, *scoring_config.get("speed_anchor_quantiles", [0.25, 0.5, 0.75]), 1]
    token_series = pl.Series("tokens", implied_token_usages)
    anchors = [
        implied_token_usages[0] if quantile == 0 else implied_token_usages[-1] if quantile == 1 else float(token_series.quantile(quantile, interpolation="linear"))
        for quantile in quantiles
    ]
    source_min = anchors[0]
    source_max = anchors[-1]
    if source_max <= source_min:
        return default_anchors
    range_min = scoring_config.get("speed_output_token_range_min", 200)
    range_max = scoring_config.get("speed_output_token_range_max", 8000)
    return [round(range_min + ((anchor - source_min) / (source_max - source_min)) * (range_max - range_min)) for anchor in anchors]


def build_scores(model: dict[str, Any], cost: Any, speed: dict[str, Any], speed_output_token_anchors: list[int], scoring_config: dict[str, Any]) -> dict[str, Any] | None:
    """Build score fields for Scoring LLM stats selection."""
    intelligence_index = _metric_value(model, "intelligence_index") or _metric_value(model, "artificial_analysis_intelligence_index")
    agentic_index = _metric_value(model, "agentic_index") or _metric_value(model, "artificial_analysis_agentic_index")
    intelligence_benchmark_mean = mean_of_finite([_metric_value(model, key) for key in scoring_config.get("intelligence_benchmark_keys", [])])
    agentic_benchmark_mean = mean_of_finite([_metric_value(model, key) for key in scoring_config.get("agentic_benchmark_keys", [])])
    intelligence_score = (intelligence_index + intelligence_benchmark_mean * 100) / 2 if intelligence_index is not None and intelligence_benchmark_mean is not None else None
    agentic_score = (agentic_index + agentic_benchmark_mean * 100) / 2 if agentic_index is not None and agentic_benchmark_mean is not None else None
    blended_price = blended_price_value(cost, scoring_config)
    latency = as_finite_number(speed.get("latency_seconds_median"))
    throughput = as_finite_number(speed.get("throughput_tokens_per_second_median"))
    e2e_latency = as_finite_number(speed.get("e2e_latency_seconds_median"))
    price_score = 1 / blended_price if blended_price is not None and blended_price > 0 else None
    imagined_speed_score = mean_of_finite(
        [target / (latency + target / throughput) if latency is not None and throughput is not None and throughput > 0 else None for target in speed_output_token_anchors]
    )
    representative_target_tokens = median(sorted(speed_output_token_anchors)) if speed_output_token_anchors else None
    observed_e2e_speed_score = representative_target_tokens / e2e_latency if representative_target_tokens is not None and e2e_latency is not None and e2e_latency > 0 else None
    speed_score = mean_of_finite([imagined_speed_score, observed_e2e_speed_score])
    if all(score is None for score in (intelligence_score, agentic_score, price_score, speed_score)):
        return None
    return {
        "intelligence_score": intelligence_score,
        "agentic_score": agentic_score,
        "speed_score": speed_score,
        "price_score": price_score,
    }


def attach_relative_scores(models: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Attach relative scores for Scoring LLM stats selection."""
    if not models:
        return []

    score_frame = pl.DataFrame(
        [
            {
                "row_index": row_index,
                "intelligence_score": as_finite_number(as_record(model.get("scores")).get("intelligence_score")),
                "agentic_score": as_finite_number(as_record(model.get("scores")).get("agentic_score")),
                "speed_score": as_finite_number(as_record(model.get("scores")).get("speed_score")),
                "price_score": as_finite_number(as_record(model.get("scores")).get("price_score")),
            }
            for row_index, model in enumerate(models)
        ]
    )

    def min_max_relative(column_name: str) -> pl.Expr:
        return (
            pl.when(pl.col(column_name).is_null() | (pl.col(column_name).count() == 0))
            .then(None)
            .when(pl.col(column_name).max() == pl.col(column_name).min())
            .then(100.0)
            .otherwise(((pl.col(column_name) - pl.col(column_name).min()) / (pl.col(column_name).max() - pl.col(column_name).min())) * 100)
            .alias(f"{column_name}_relative")
        )

    def percentile_relative(column_name: str) -> pl.Expr:
        return (
            pl.when(pl.col(column_name).is_null() | (pl.col(column_name).count() == 0))
            .then(None)
            .otherwise((pl.col(column_name).rank(method="max") / pl.col(column_name).count()) * 100)
            .alias(f"{column_name}_relative")
        )

    relative_frame = score_frame.with_columns(
        [
            min_max_relative("intelligence_score"),
            min_max_relative("agentic_score"),
            percentile_relative("speed_score"),
            percentile_relative("price_score"),
        ]
    ).with_columns(
        pl.mean_horizontal(
            "intelligence_score_relative",
            "agentic_score_relative",
            "speed_score_relative",
            "price_score_relative",
        ).alias("overall_score_relative")
    )

    relative_by_index = {
        row["row_index"]: row
        for row in relative_frame.select(
            "row_index",
            "intelligence_score_relative",
            "agentic_score_relative",
            "speed_score_relative",
            "price_score_relative",
            "overall_score_relative",
        ).to_dicts()
    }
    return [
        {
            **model,
            "relative_scores": {
                "intelligence_score": relative_by_index[row_index]["intelligence_score_relative"],
                "agentic_score": relative_by_index[row_index]["agentic_score_relative"],
                "speed_score": relative_by_index[row_index]["speed_score_relative"],
                "price_score": relative_by_index[row_index]["price_score_relative"],
                "overall_score": relative_by_index[row_index]["overall_score_relative"],
            },
        }
        for row_index, model in enumerate(models)
    ]

File: llm-harness-js/src/stats/llm/llm-stats/scoring.ts
import { percentileRank } from "../../utils.js";
import { asFiniteNumber, asRecord, type JsonObject } from "../shared.js";

import { type ModelStatsSelectedModel } from "./types.js";

const INTELLIGENCE_BENCHMARK_KEYS = [
  "omniscience_accuracy",
  "hle",
  "lcr",
  "scicode",
] as const;
const AGENTIC_BENCHMARK_KEYS = [
  "omniscience_nonhallucination_rate",
  "gdpval_normalized",
  "ifbench",
  "terminalbench_hard",
] as const;
const DEFAULT_SPEED_OUTPUT_TOKEN_ANCHORS = [
  200, 500, 1_000, 2_000, 8_000,
] as const;
const SPEED_OUTPUT_TOKEN_RANGE_MIN = 200;
const SPEED_OUTPUT_TOKEN_RANGE_MAX = 8_000;
const SPEED_ANCHOR_QUANTILES = [0.25, 0.5, 0.75] as const;
const WEIGHTED_PRICE_INPUT_RATIO = 0.75;
const WEIGHTED_PRICE_OUTPUT_RATIO = 0.25;

function meanOfFinite(values: Array<number | null>): number | null {
  const finiteValues = values.filter(
    (value): value is number => value != null && Number.isFinite(value),
  );
  if (finiteValues.length === 0) {
    return null;
  }
  const total = finiteValues.reduce((sum, value) => sum + value, 0);
  return total / finiteValues.length;
}

function metricValue(model: JsonObject, key: string): number | null {
  const intelligence = asRecord(model.intelligence);
  const evaluations = asRecord(model.evaluations);
  if (key === "omniscience_nonhallucination_rate") {
    const nonhallucinationRate = asFiniteNumber(
      intelligence.omniscience_nonhallucination_rate,
    );
    if (nonhallucinationRate != null) {
      return nonhallucinationRate;
    }
    const nonhallucinationRateFromLegacyKey = asFiniteNumber(
      intelligence.omniscience_hallucination_rate,
    );
    return nonhallucinationRateFromLegacyKey;
  }
  return (
    asFiniteNumber(intelligence[key]) ??
    asFiniteNumber(evaluations[key]) ??
    null
  );
}

export function blendedPriceValue(costLike: unknown): number | null {
  const cost = asRecord(costLike);
  const inputCost = asFiniteNumber(cost.input);
  const outputCost = asFiniteNumber(cost.output);
  const weightedInputCost = asFiniteNumber(cost.weighted_input);
  const weightedOutputCost = asFiniteNumber(cost.weighted_output);
  const cacheReadCost = asFiniteNumber(cost.cache_read);
  const cacheWriteCost = asFiniteNumber(cost.cache_write);
  if (
    inputCost == null ||
    outputCost == null ||
    inputCost <= 0 ||
    outputCost <= 0
  ) {
    return null;
  }
  if (weightedInputCost != null || weightedOutputCost != null) {
    const effectiveInputCost =
      weightedInputCost != null ? weightedInputCost : inputCost;
    const effectiveOutputCost =
      weightedOutputCost != null ? weightedOutputCost : outputCost;
    return (
      WEIGHTED_PRICE_INPUT_RATIO * effectiveInputCost +
      WEIGHTED_PRICE_OUTPUT_RATIO * effectiveOutputCost
    );
  }

  const cacheWeightedInput = cacheReadCost != null ? cacheReadCost : inputCost;
  const cacheWeightedOutput =
    cacheWriteCost != null
      ? 0.1 * cacheWriteCost + 0.9 * outputCost
      : outputCost;
  const baseProxy =
    WEIGHTED_PRICE_INPUT_RATIO *
      (WEIGHTED_PRICE_INPUT_RATIO * cacheWeightedInput +
        WEIGHTED_PRICE_OUTPUT_RATIO * inputCost) +
    WEIGHTED_PRICE_OUTPUT_RATIO * cacheWeightedOutput;

  const over200kCost = asRecord(cost.context_over_200k);
  const over200kInput = asFiniteNumber(over200kCost.input);
  const over200kOutput = asFiniteNumber(over200kCost.output);
  const over200kCacheRead = asFiniteNumber(over200kCost.cache_read);
  const over200kCacheWrite = asFiniteNumber(over200kCost.cache_write);
  if (
    over200kInput == null ||
    over200kOutput == null ||
    over200kInput <= 0 ||
    over200kOutput <= 0
  ) {
    return baseProxy;
  }

  const over200kInputWeighted =
    over200kCacheRead != null ? over200kCacheRead : over200kInput;
  const over200kOutputWeighted =
    over200kCacheWrite != null
      ? 0.1 * over200kCacheWrite + 0.9 * over200kOutput
      : over200kOutput;
  const over200kProxy =
    WEIGHTED_PRICE_INPUT_RATIO *
      (WEIGHTED_PRICE_INPUT_RATIO * over200kInputWeighted +
        WEIGHTED_PRICE_OUTPUT_RATIO * over200kInput) +
    WEIGHTED_PRICE_OUTPUT_RATIO * over200kOutputWeighted;

  return 0.95 * baseProxy + 0.05 * over200kProxy;
}

function quantileFromSorted(values: number[], quantile: number): number | null {
  if (values.length === 0) {
    return null;
  }
  if (values.length === 1) {
    return values[0] ?? null;
  }
  const clampedQuantile = Math.min(1, Math.max(0, quantile));
  const index = (values.length - 1) * clampedQuantile;
  const lowerIndex = Math.floor(index);
  const upperIndex = Math.ceil(index);
  if (lowerIndex === upperIndex) {
    return values[lowerIndex] ?? null;
  }
  const lowerValue = values[lowerIndex];
  const upperValue = values[upperIndex];
  if (lowerValue == null || upperValue == null) {
    return null;
  }
  const ratio = index - lowerIndex;
  return lowerValue + (upperValue - lowerValue) * ratio;
}

export function deriveSpeedOutputTokenAnchors(
  openRouterSpeedById: Map<string, JsonObject>,
): number[] {
  const impliedTokenUsages = Array.from(openRouterSpeedById.values())
    .map((speed) => {
      const throughputTokensPerSecond = asFiniteNumber(
        speed.throughput_tokens_per_second_median,
      );
      const latencySeconds = asFiniteNumber(speed.latency_seconds_median);
      const e2eLatencySeconds = asFiniteNumber(
        speed.e2e_latency_seconds_median,
      );
      if (
        throughputTokensPerSecond == null ||
        throughputTokensPerSecond <= 0 ||
        latencySeconds == null ||
        e2eLatencySeconds == null
      ) {
        return null;
      }
      const generationSeconds = e2eLatencySeconds - latencySeconds;
      if (generationSeconds <= 0) {
        return null;
      }
      return generationSeconds * throughputTokensPerSecond;
    })
    .filter((value): value is number => value != null && Number.isFinite(value))
    .sort((left, right) => left - right);

  if (impliedTokenUsages.length === 0) {
    return [...DEFAULT_SPEED_OUTPUT_TOKEN_ANCHORS];
  }

  const q0 = impliedTokenUsages[0] ?? null;
  const [q1, q2, q3] = SPEED_ANCHOR_QUANTILES.map((quantile) =>
    quantileFromSorted(impliedTokenUsages, quantile),
  );
  const q4 = impliedTokenUsages[impliedTokenUsages.length - 1] ?? null;
  const numericQuantileAnchors = [q0, q1, q2, q3, q4].filter(
    (value): value is number => value != null && Number.isFinite(value),
  );
  if (numericQuantileAnchors.length !== 5) {
    return [...DEFAULT_SPEED_OUTPUT_TOKEN_ANCHORS];
  }

  const sourceMin = numericQuantileAnchors[0] as number;
  const sourceMax = numericQuantileAnchors[
    numericQuantileAnchors.length - 1
  ] as number;
  if (!(sourceMax > sourceMin)) {
    return [...DEFAULT_SPEED_OUTPUT_TOKEN_ANCHORS];
  }

  return numericQuantileAnchors.map((anchor) => {
    const normalized = (anchor - sourceMin) / (sourceMax - sourceMin);
    const mapped =
      SPEED_OUTPUT_TOKEN_RANGE_MIN +
      normalized *
        (SPEED_OUTPUT_TOKEN_RANGE_MAX - SPEED_OUTPUT_TOKEN_RANGE_MIN);
    return Math.round(mapped);
  });
}

export function buildScores(
  model: JsonObject,
  cost: unknown,
  speed: JsonObject,
  speedOutputTokenAnchors: number[],
): unknown {
  const intelligenceIndex =
    metricValue(model, "intelligence_index") ??
    metricValue(model, "artificial_analysis_intelligence_index");
  const agenticIndex =
    metricValue(model, "agentic_index") ??
    metricValue(model, "artificial_analysis_agentic_index");
  const intelligenceBenchmarkMean = meanOfFinite(
    INTELLIGENCE_BENCHMARK_KEYS.map((key) => metricValue(model, key)),
  );
  const agenticBenchmarkMean = meanOfFinite(
    AGENTIC_BENCHMARK_KEYS.map((key) => metricValue(model, key)),
  );
  const intelligenceScore =
    intelligenceIndex != null && intelligenceBenchmarkMean != null
      ? (intelligenceIndex + intelligenceBenchmarkMean) / 2
      : null;
  const agenticScore =
    agenticIndex != null && agenticBenchmarkMean != null
      ? (agenticIndex + agenticBenchmarkMean) / 2
      : null;
  const blendedPrice = blendedPriceValue(cost);
  const latencySeconds = asFiniteNumber(speed.latency_seconds_median);
  const throughputTokensPerSecond = asFiniteNumber(
    speed.throughput_tokens_per_second_median,
  );
  const e2eLatencySeconds = asFiniteNumber(speed.e2e_latency_seconds_median);
  const priceScore =
    blendedPrice != null && blendedPrice > 0 ? 1 / blendedPrice : null;
  const imaginedSpeedScore = meanOfFinite(
    speedOutputTokenAnchors.map((targetTokens) =>
      latencySeconds != null &&
      throughputTokensPerSecond != null &&
      throughputTokensPerSecond > 0
        ? targetTokens /
          (latencySeconds + targetTokens / throughputTokensPerSecond)
        : null,
    ),
  );
  const sortedAnchors = [...speedOutputTokenAnchors].sort(
    (left, right) => left - right,
  );
  const representativeTargetTokens = quantileFromSorted(sortedAnchors, 0.5);
  const observedE2eSpeedScore =
    representativeTargetTokens != null &&
    e2eLatencySeconds != null &&
    e2eLatencySeconds > 0
      ? representativeTargetTokens / e2eLatencySeconds
      : null;
  const speedScore = meanOfFinite([imaginedSpeedScore, observedE2eSpeedScore]);
  if (
    intelligenceScore == null &&
    agenticScore == null &&
    priceScore == null &&
    speedScore == null
  ) {
    return null;
  }
  return {
    intelligence_score: intelligenceScore,
    agentic_score: agenticScore,
    speed_score: speedScore,
    price_score: priceScore,
  };
}

export function attachPercentiles(
  models: ModelStatsSelectedModel[],
): ModelStatsSelectedModel[] {
  const intelligenceScores = models.map((model) =>
    asFiniteNumber(asRecord(model.scores).intelligence_score),
  );
  const agenticScores = models.map((model) =>
    asFiniteNumber(asRecord(model.scores).agentic_score),
  );
  const speedScores = models.map((model) =>
    asFiniteNumber(asRecord(model.scores).speed_score),
  );
  const priceScores = models.map((model) =>
    asFiniteNumber(asRecord(model.scores).price_score),
  );

  return models.map((model) => {
    const scores = asRecord(model.scores);
    const intelligenceScore = asFiniteNumber(scores.intelligence_score);
    const agenticScore = asFiniteNumber(scores.agentic_score);
    const speedScore = asFiniteNumber(scores.speed_score);
    const priceScore = asFiniteNumber(scores.price_score);
    const intelligencePercentile =
      intelligenceScore == null
        ? null
        : percentileRank(intelligenceScores, intelligenceScore);
    const agenticPercentile =
      agenticScore == null ? null : percentileRank(agenticScores, agenticScore);
    const speedPercentile =
      speedScore == null ? null : percentileRank(speedScores, speedScore);
    const pricePercentile =
      priceScore == null ? null : percentileRank(priceScores, priceScore);
    return {
      ...model,
      percentiles: {
        intelligence_percentile: intelligencePercentile,
        agentic_percentile: agenticPercentile,
        speed_percentile: speedPercentile,
        price_percentile: pricePercentile,
      },
    };
  });
}
